### PR TITLE
Add lint for maps with zero-sized value types

### LINKED ISSUE #1641: Lint `HashMap<T, ZST>` where ZST is any zero sized type
The user should be using `HashSet<T>` instead and magically create the ZSTs out of thin air when needed (since they are all the same value anyway).

### PR DESCRIPTION
**Pull request by korrat**
Hi, this is my first time contributing to clippy or rust in general, so I'm not sure about the details of contributing. Please excuse me and let me now if I did anything wrong. I have a couple of questions:

1. I'm not sure what category this lint should be. I've put it in "nursery" for now.
1. Should I squash commits this is reviewed/merged?

changelog: Add lint for maps with zero-sized value types 

Fixes #1641

### CODE REVIEW

**Comment by Manishearth** [comment]
I'm in the middle of a move and may cut back on reviewing for a few weeks. I'm going to reassign my PRs for now, though i'll still try to review smaller things.

r? @ebroto

File: clippy_lints/src/zero_sized_map_values.rs
Comment (ebroto) [inline clippy_lints/src/zero_sized_map_values.rs:36]:
```suggestion
    complexity,
```

File: tests/ui/zero_sized_map_values.rs
Comment (ebroto) [inline tests/ui/zero_sized_map_values.rs:1]: Let's add a couple more tests using
* `static`
* `const`
* type alias
* `enum`
* a method

File: clippy_lints/src/zero_sized_map_values.rs
Comment (ebroto) [inline clippy_lints/src/zero_sized_map_values.rs:19]: As highlighted in [this comment](https://github.com/rust-lang/rust-clippy/pull/6165#pullrequestreview-510993250), we should not lint in public item signatures, or leave that for a follow-up and mention that in the `Known problems` section.

If we go for the latter, then the category should be `pedantic` instead of `complexity`.

**Review by ebroto — CHANGES_REQUESTED** [review]
Thanks for your contribution and sorry for the delay in reviewing, I'm really short on time lately.

---

I think we can simplify a lot this lint if we move the body of your current `check_ty` method to the [implementation of LateLintPass' `check_ty`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/passes/trait.LateLintPass.html#method.check_ty), transforming the `hir::Ty` into a `ty::Ty` with `hir_ty_to_ty` as you do currently elsewhere. 

We should keep the check you do to avoid linting in trait impls, that is a good idea :+1:

**Comment by ebroto** [comment]
>Should I squash commits this is reviewed/merged?

In this case, I think it makes sense as there are some intermediary commits. But let's wait until the PR is approved, it's more helpful if the remarks are addressed as follow-up commits.

File: clippy_lints/src/zero_sized_map_values.rs
Comment (korrat) [reply clippy_lints/src/zero_sized_map_values.rs]: Moved most of the implementation into `LateLintPass::check_ty`

File: tests/ui/zero_sized_map_values.rs
Comment (korrat) [reply tests/ui/zero_sized_map_values.rs:1]: Added tests for these items

File: clippy_lints/src/zero_sized_map_values.rs
Comment (korrat) [reply clippy_lints/src/zero_sized_map_values.rs]: * Updated known problems
* Changed category to pedantic

**Comment by korrat** [comment]
This PR currently causes some tests to fail with ICEs. I believe that this is due to the use of `hir_ty_to_ty` which returns `TyKind::Error` in some cases. Commenting out this part of the code makes the tests pass again. So far I have been unable to find a way around the use of `hir_ty_to_ty` without severely limiting the scope of the analysis. 

Is there any way to turn a HIR `Ty` into a `Ty` with explicit error handling? Otherwise, I see the following options: 

1. Implement the lint only in local scopes for now. 
2. Only consider the obvious case of `HashMap<K, ()>` when checking non-local items

What is your opinion on this?

**Review by ebroto — COMMENTED** [review]
Indeed, `hir_ty_to_ty` seems to choke with inference variables and unnamed lifetimes. To solve this, the proposed solution first would check if there are typecheck results available and use those, otherwise fail back to `hir_ty_to_ty`. 

Typecheck results are available in bodies, where inferred types are valid, so if there are no results we should not find type inference.

**Comment by giraffate** [comment]
ping from triage @korrat. There seems to be fixes left to be done. Do you have any questions on how to proceed here?

File: .cargo/config
Comment (korrat) [reply .cargo/config]: Sorry, that shouldn't have been added. I was fighting sccache.

File: tests/ui/zero_sized_map_values.rs
Comment (korrat) [reply tests/ui/zero_sized_map_values.rs:1]: Added a test case with BTreeMap. Should I add additional tests for other cases too or do we trust that BTreeMap and HashMap are treated equally?

Comment (ebroto) [reply tests/ui/zero_sized_map_values.rs:1]: I think it would be preferable to check all the cases for both containers.

A simple way to go would be to copy the current test file and change the `HashMap`s to `BTreeMap`s. We can name the files `zero_sized_hashmap_values.rs` and `zero_sized_btreemap_values.rs` for example.

File: tests/ui/zero_sized_map_values.rs
Comment (ebroto) [inline tests/ui/zero_sized_map_values.rs:64]: we can remove this line as we have much more tests than these :)

**Review by ebroto — CHANGES_REQUESTED** [review]
Just the nit (sorry about that :smile:) and your suggestion about the tests, and this should be ready to go!

Thanks for making this move forward.

**Comment by korrat** [comment]
Thanks for being so quick in reviewing this again and again and sorry for the delay previously. Somehow the notification about your comment ended up in my spam mail folder where I didn't see it until yesterday. 

@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

**Review by ebroto — APPROVED** [review]
This LGTM now! Thanks for all the effort you've put into this, and don't hesitate to take another issue if you had fun with this one :)

Could you squash your commits before merging? I should have said that in my last review, my apologies.

**Comment by korrat** [comment]
Sure, I squashed the commits. I didn't want to make the review harder for you earlier.

**Comment by ebroto** [comment]
@bors r+

Thanks!

**Pull request merged**
